daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -539,39 +471,57 @@ object ContractStateMachine {
-    *     excluding nodes that have been rolled back.
-    *
-    *   - In mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]]
-    *     the following operations mutate this map:
-    *     1. fetch-by-key/lookup-by-key/exercise-by-key/create-contract-with-key will insert an
-    *        an entry in the map if there wasn’t already one (i.e., if they queried the ledger).
-    *     2. ACS mutating operations if the corresponding contract has a key update the entry. Specifically,
-    *        2.1. A create will set the corresponding map entry to KeyActive(cid) if the contract has a key.
-    *        2.2. A consuming exercise on cid will set the corresponding map entry to KeyInactive
-    *             iff we had a KeyActive(cid) entry for the same key before. If not, keys
-    *             will not be modified.
-    *             Later lookups have an activeness check
-    *             that can then set this to KeyInactive if the result of the
-    *             lookup was already archived.
+    * @param localKeys
+    *   A store of the latest local contract that has been created with the given key in this timeline.
+    *   Later creates overwrite earlier ones. Note that this does not track whether the contract
+    *   was consumed or not. That information is stored in consumedBy.
+    *   It also _only_ includes local contracts not global contracts.
     */
   final case class ActiveLedgerState[+Nid](
       locallyCreatedThisTimeline: Set[ContractId],
       consumedBy: Map[ContractId, Nid],
-      keys: Map[GlobalKey, KeyMapping],
+      private val localKeys: Map[GlobalKey, Value.ContractId],
   ) {
     def consume[Nid2 >: Nid](contractId: ContractId, nodeId: Nid2): ActiveLedgerState[Nid2] =
       this.copy(consumedBy = consumedBy.updated(contractId, nodeId))
+
+    def createKey(key: GlobalKey, cid: Value.ContractId): ActiveLedgerState[Nid] =
+      this.copy(localKeys = localKeys.updated(key, cid))
+
+    /** Equivalence relative to locallyCreatedThisTimeline, consumedBy & localActiveKeys.
+      */
+    def isEquivalent[Nid2 >: Nid](other: ActiveLedgerState[Nid2]): Boolean =
+      this.locallyCreatedThisTimeline == other.locallyCreatedThisTimeline &&
+        this.consumedBy == other.consumedBy &&
+        this.localActiveKeys == other.localActiveKeys
+
+    /** See docs of [[ContractStateMachine.advance]]
+      */
+    private[ContractStateMachine] def advance[Nid2 >: Nid](
+        substate: ActiveLedgerState[Nid2]
+    ): ActiveLedgerState[Nid2] =
+      ActiveLedgerState(
+        locallyCreatedThisTimeline = this.locallyCreatedThisTimeline
+          .union(substate.locallyCreatedThisTimeline),
+        consumedBy = this.consumedBy ++ substate.consumedBy,
+        localKeys = this.localKeys.concat(substate.localKeys),
+      )
+
+    /** localKeys filter by whether contracts have been consumed already.
+      */
+    def localActiveKeys: Map[GlobalKey, KeyMapping] =
```

Comment:
Note that I had to slightly change localActiveKeys here. It doesn’t drop entries but only sets them to `Inactive`.

The reason for that is our multi-key semantics:
1. Global lookup resolves against cid 1.
2.  Local create with cid 2
3. Archive of cid 2
4. Currently, we now resolve against Keyinactive not cid 1.


Now my proposal for multi-key semantics changes this but I don’t want to change this in this PR.

Reply:
Yeah, let's keep the non-UCK semantics stable for now.